### DISCUSSION TITLE
Negative value for positional argument with command chaining

### DISCUSSION THREAD

**Discussion by abey79** [pacak/bpaf#356] [Q&A] [answered]
Considering this modified command chaining example:
```rust
use bpaf::{positional, Bpaf, Parser};

#[derive(Debug, Clone, Bpaf)]
#[bpaf(options)]
pub struct Options {
    #[bpaf(external(cmd), many)]
    pub commands: Vec<Cmd>,
}

#[derive(Debug, Clone, Bpaf)]
pub enum Cmd {
    #[bpaf(command, adjacent)]
    /// Performs drinking action
    Drink {
        #[bpaf(positional("BEVERAGE"))]
        coffee: isize,  // <------------- KEY CHANGE HERE
    },

    #[bpaf(command, adjacent)]
    /// Performs taking a nap action
    Sleep,
}

fn main() {
    println!("{:?}", options().run())
}
```

I can't find a way to allow passing negative values to the `drink` command. With positional argument, neither of the documented was seem possible. Using `=` with positional arg isn't possible, and `--` doesn't work either:

```
$ negative_positional drink -- -1 
Error: expected BEVERAGE, pass --help for usage information
```

Is there another way to support positional negative values?

**Comment by pacak** [comment, marked as answer]
> Is there another way to support positional negative values?

https://github.com/pacak/bpaf/discussions/347

**Reply by abey79** [reply]
Indeed I could make it work this way—sorry for overlooking that question and thanks for the super-fast answers!

That said, given that I'm working on a numerically-heavy CLI (lots of coordinates, transforms, etc.—for context, I'm working on a rust port of [this python project](https://vpype.readthedocs.io/en/latest/reference.html)), I would need to have these `any::<_>("META", Some)` ~~`, anywhere`~~ attribute~~s~~ absolutely all over the place.

Would you be open to consider adding some helper to reduce this boilerplate?

```diff
use bpaf::{Bpaf, Parser};

#[derive(Debug, Clone, Bpaf)]
#[bpaf(options)]
pub struct Options {
    #[bpaf(external(cmd), many)]
    pub commands: Vec<Cmd>,
}

#[derive(Debug, Clone, Bpaf)]
pub enum Cmd {
    #[bpaf(command, adjacent)]
    /// Performs drinking action
    Drink {
-       #[bpaf(positional("BEVERAGE"))]
+       #[bpaf(any::<_>("BEVERAGE", Some))]
        coffee: isize,
    },

    #[bpaf(command, adjacent)]
    /// Performs taking a nap action
    Sleep,
}

fn main() {
    println!("{:?}", options().run())
}
```

**Reply by pacak** [reply]
`anywhere` means it behaves like a flag - probably not what you want to deal with coordinates or transformations. If you have lots of coordinates of offsets - I'd define them as top level primitives  (`fn x_coordinate() -> impl Parser<isize> {..}`) and call it with externals. Alternatively I think all you need is `#[bpaf(any::<isize>("BEVERAGE", Some))]`.  I have plans to allow for slightly more flexible version of `external`, but nothing yet implemented.

**Reply by abey79** [reply]
- Thanks for the note on `anywhere`, indeed I don't need it to make it work (will update the snippet above for future reference).
- I tend to have different METAVAR everywhere, so I'd indeed need to have support for something like `#[bpaf(external(numeric("META")))]` (which is still a lot of boilerplate :) )
- It would also be OK for me if `--` worked as expected (aka separator between named and positional args _for the current command_) in chained command context. FWIW, that's how [Click](https://click.palletsprojects.com/en/8.1.x/) addresses that issue.

**Reply by abey79** [reply]
Side note: the type can be inferred, avoiding repeating it in the `any` attribute.

**Reply by abey79** [reply]
`positional("META", any)` could be nice syntactic sugar for `any::<_>("META", Some)`?

**Reply by pacak** [reply]
> Side note: the type can be inferred, avoiding repeating it in the any attribute.

`any` is an escape hatch that lets you parse strange things rather than something you would use normally, type cannot be inferred, at least not with information available in a proc macro. Currently `any` takes a function and that function can change a type `any` produces.

I'm trying to keep combinatorial and derive APIs close to each other and `positional("META", any)` breaks this principle for not that much gain.

> It would also be OK for me if `--` worked as expected (aka separator between named and positional args for the current command) in chained command context.

This is an interesting thought, I'll see how far I can get it.

**Reply by abey79** [reply]
>  type cannot be inferred

I mean, it _does_ work with `any::<_>("BEVERAGE", Some)`, so type is inferred by rustc (but indeed not by the proc macro).
